### DISCUSSION TITLE
Using ScriptKit to Automate Mac Setup

### DISCUSSION THREAD

**Discussion by jh3y** [johnlindquist/kit#71] [Share]
Howdy y'all! ＼ʕ •ᴥ•ʔ／

I decided I'd have one of those "tech" weekends where I get all my tech in order 😅  A big part of this was getting my machines onto `Big Sur` with a blank slate and revisiting what things I install etc.

For the last few years, I've been using [Kody](https://github.com/jh3y/kody) which is something I built with Node. It's one of those "If it ain't broke, don't fix it" kinda things. It works great but once I saw "ScriptKit", that was kinda the thing I envisioned those years ago. The GUI part is a bonus!

Enough ramble. How did it go? What's my process?

1. Fresh Big Sur install on clean machine.
2. Install ScriptKit
3. That prompts Git installation 👍 
4. Install HomeBrew
5. Install VSCode (Doing this so I can mess about with scripts before running them). Imagine there would be a way to automate the steps from 4 onwards.
6. Use Terminal to add a remote for my scripts.
7. Switch to that branch.
8. This part was a bit of a sticker. I want to run my setup scripts (there a little rough, don't judge! 😅 ) from the CLI initially so I can see what's happening. But, I'm not version controlling the `bin` files. Perhaps there is an automated way for me to do this. Who am I kidding? Of course there is, this is ScriptKit. But, I didn't know if there was an API I could tap into to say "Hey! Here's my scripts I brought with me. Can you hook them up to bin please?"
8a. Make a couple of tweaks to `cli`. Expose `basename` from `path` and `rm` from `shelljs`. Expose `iterm` as a global for CodePen script.
9. Then I do a bit of a dance with setting the path in `.zshrc` 😅 
10. Run each of my scripts. One installs all my Homebrew packages from a config file, another symlinks all my dotfiles(which I'm currently keeping in a folder within scripts 😅 ). That sets up my VSCode, sets up my Git Config, ZSH, etc. Then I have another script that runs a shell script with my OSX defaults like hide the dock, set up exposé, etc. And lastly, a script that installs global npm modules, installs VSCode extensions, etc.
11. Lastly, grab my iTerm config and Stream Deck backups from within the "scripts" folder and apply them too.
12. 🎉  Machine ready for action.

I'm thrilled it works. My scripts are a little "rough" around the edges 😅 

But, as a POC, it works! And this saved me hours of doing this myself and remembering things I install or set up. Migrating what I had already wasn't much hassle at all.

I've had some thoughts too about ways I could make this easier to sync what I'm using too.

- Make an install script that updates my config for me upon successful HomeBrew package install 🤩  Same with uninstall.
- Script to grab iTerm profile for me and copy it into my branch.

Another note.
The file selector didn't work for me still 🤔  I'm not sure what's going on there.

This was great fun to put together and was totally worth spending the time on to get my tech in order 💪 

Thanks @johnlindquist 🙏  Keep up the great work!

**Comment by johnlindquist** [comment]
Thank you so much for sharing this!

One of the core values of Script Kit is "scripts can run in the terminal and in the App". The ~/.kit has just as much as future running in your CI as it does on your desktop. In fact, the [CI in this repo](https://github.com/johnlindquist/kit/blob/main/.github/workflows/test.yml#L90) uses ~/.kit to test itself.

Thoughts on helping your setup process.:
1. The ~/.kenv is intended for App scripts. I would create a separate ~/.setup-kenv to store setup scripts that will only be run from the terminal. Note: All kenvs leverage the same ~/.kit
2. The [~/.kit/script](https://github.com/johnlindquist/kit/blob/main/script) executable can be pointed at any .js file on your system. So `~/.kit/script ~/.kenv/scripts/install-ffmpeg.js` totally works. (Or should, I haven't tested in a while)
3. There's a [add-kenv-to-profile](https://github.com/johnlindquist/kit/blob/main/cli/add-kenv-to-profile.js) script which can take care of adding to your bash (it's checks for zsh, bash, fish)
4. There's also a [create-bin](https://github.com/johnlindquist/kit/blob/main/cli/create-bin.js) script which you can run against any existing script. This script is the one that runs anytime you run "new"

**Reply by jh3y** [reply]
Love that @johnlindquist 

Thank you 🙏  A bunch of actionable things I can take and do with that.

1. How do you hook up `kit` to look at that separate `.kenv` too? Is there a way to say, "Grab scripts from here too!"?
2. This is rad. I'll check that out.
3. Yeah, I saw this. I should give it another go. I don't think it did but I need to make sure it doesn't wipe out my HomeBrew path at the same time. Also, when I symlink my `.zshrc`, then I have the exported path to ScriptKit in there now.
4. Ahh, awesome. That would be useful. I could perhaps create a little loop or something. That would sort that out.

I'm still not sure why I can't use the file selector. But, that's another issue I imagine. I'd love to have the script prompt me to select an image file and then use it for my OG generator. But, it doesn't seem to do anything. All permissions are allowed, etc. 🤔

**Reply by johnlindquist** [reply]
> 1. How do you hook up kit to look at that separate .kenv too? Is there a way to say, "Grab scripts from here too!"?

A. The Kit App only looks to the main .kenv environment. If I add the ability for it to use other kenvs, then it would be a "switch environment", not a "pull from multiple environments". Each environment has its own .env which could cause clashes.

B. 
1. The "heart" of Script Kit lies in [this template](https://github.com/johnlindquist/kit/blob/main/templates/bin/template). I need make a video on what's going on here, but it's essentially linking all the necessary bits together. 
2. The [env template](https://github.com/johnlindquist/kit/blob/main/templates/env/template.env) is also important and more straightforward.
3. After cloning a new kenv, running this will link those 2 templates above together (notice the `KENV` env var):
```bash 
KENV=~/.my-new-kenv ~/.kit/script ~/.kit/setup/setup.js
```

> 3. I need to make sure it doesn't wipe out my HomeBrew path
The script uses "append to file" logic, so it doesn't touch any contents of the file

> 4. a little loop
Yup, it's just "a script". Here's the setup version which does exactly that: [create-cli-bins](https://github.com/johnlindquist/kit/blob/main/setup/create-cli-bins.js)
